2단계 - 지하철 구간 제거 기능 개선

road-jin opened the pull request:
안녕하세요 홍준님~
이번 미션도 인수테스트는 성공 케이스, 도메인 테스트에서는 실패 케이스만 작성하였습니다.

Sections가 코드양이 많아서 추가, 삭제를 위임할 클래스(SectionsAppender, SectionsRemover) 만들어야 할지 고민이 됩니다.
홍준님은 클래스를 분리할 때 어떤 기준으로 분리하시나요?

road-jin committed:
fix: 지하철 구간 삭제 수정

구간 삭제시 중간 구간을 삭제할 수 있게 수정

hjhello423 commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -83,26 +79,51 @@ private int calculateDistance(int originalDistance, int newDistance) {
+	private void removeSectionWhenSectionNotEmpty(Optional<Section> upSection, Optional<Section> downSection) {
+		upSection.ifPresent(s -> sections.remove(s));
+		downSection.ifPresent(s -> sections.remove(s));
+	}
```
이 메서드는 Optional을 인자로 받고 있네요. 🤔
Optional에 대해 찾아보고 리팩토링해보면 좋을 것 같습니다!

hjhello423 commented on src/test/java/nextstep/subway/acceptance/SectionAcceptanceTest.java:
```diff
@@ -111,34 +111,21 @@ void init() {
 	 */
 	@Test
-	void 지하철_노선에_구간을_삭제할_떄_마지막_구간이_아닌_경우_실패한다() {
+	void 지하철_노선에_중간_구간을_삭제한다() {
```
미션의 기능 요구 사항들에 대한 테스트가 부족해 보이는데요.
학습 과정인 만큼 추가해 보고 이번 step을 마무리해 보면 좋을 것 같아요. 😀
추가된 기능 들의 정상 동작과 예외 케이스에 대한 인수 테스트를 추가해 주세요!

hjhello423 commented on src/test/java/nextstep/subway/unit/SectionsTest.java:
```diff
@@ -54,4 +55,11 @@ void setUp() {
 			() -> sections.addSection(new Section(신분당선, 강남역, 양재역, 5)));
 	}
+
+	@Test
+	void 구간을_삭제할_때_마지막_구간이면_에러를_반환한다() {
```
`Sections` 도메인에서도 변경사항이 많았지만 그에 비해 테스트는 부족해 보여요.

hjhello423 commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -12,10 +13,9 @@
-import nextstep.subway.exception.SectionRemoveSizeException;
 
 @Embeddable
 public class Sections {
```
> Sections가 코드양이 많아서 추가, 삭제를 위임할 클래스(SectionsAppender, SectionsRemover) 만들어야 할지 고민이 됩니다.
홍준님은 클래스를 분리할 때 어떤 기준으로 분리하시나요?

어려운 질문이네요. 
우선 당연하게도 상황에 따라 다를 것 같고, 기능들은 도메인 베이스로 위치시키고 있고요.
중요한 비즈니스 기능이 있는 경우엔 분리를 시도해 볼 수 있을 것 같네요.
다양한 디자인 패턴들이 있는데 [참고하면 좋을 사이트](https://refactoring.guru/ko/design-patterns) 하나 공유드립니다!

hjhello423 reviewed (changes requested):
안녕하세요 민규님 😀
인수 테스트를 학습하는 과정인 만큼 테스트 케이스들을 추가해 보면 좋을 것 같습니다!
코멘트 확인해 보시고 다시 한번 리뷰 요청 주세요.

road-jin committed:
refactor: 구간 삭제 인수 테스트 피드백 반역

구간 삭제시 노선에 등록 되어 있지 않은 역일 경우 예외 발생하게 추가
구간 추가 및 삭제 실패 인수 테스트 추가
구간 추가 및 삭제 성공 도메인 단위 테스트 추가

road-jin commented:
홍준님 피드백 감사합니다 ㅎㅎ

인수 테스트에 예외 케이스 추가하고, 도메인 단위 테스트에는 성공 케이스 추가하였습니다.
그리고 Optional 리팩토링을 진행하라고 하셔서 찾아보았는데,
파라미터로 Optional를 사용할 경우
Optional 객체가 null인 경우, Optional로 감싸진 제너릭 클래스가 null인 경우 두번을 체크해야 되서 
오히려 안 좋다는 것을 확인하였습니다.
의도하신 바가 맞은신지 궁굼합니다.

공유해주신 사이트는 차근차근 공부해보겠습니다.
감사합니다.

hjhello423 commented on src/test/java/nextstep/subway/acceptance/SectionAcceptanceTest.java:
```diff
@@ -92,50 +96,128 @@ void init() {
+		지하철_노선의_역들을_조회하여_변경된_구간을_검증(신분당선.getId(), 신사역, 논현역);
+	}
+
+	/**
+	 * Given 자하철 노선에 구간을 포함하여 생성하고
+	 * When 지하철 노선에 마지막 구간을 삭제하면
+	 * Then 요청이 실패된다
+	 */
+	@Test
+	void 구간을_삭제할_때_마지막_구간이면_요청이_실패된다() {
+		// when
+		var response = 지하철_노선의_구간_삭제_요청(신분당선.getId(), 신논현역.getId());
 
 		// then
 		응답_상태코드_검증(response, HttpStatus.BAD_REQUEST);
 	}
```
시나리오와 테스트가 일치하지 않는 것 같아요. 😭
코드는 등록되지 않은 역을 삭제하고 있는 것으로 보이네요.

road-jin replied:
앗 그러네요... 이건 다음 요청 때 바로 반영하겠습니다.

hjhello423 commented on src/test/java/nextstep/subway/unit/SectionsTest.java:
```diff
@@ -1,5 +1,7 @@
 package nextstep.subway.unit;
 
+import static org.assertj.core.api.Assertions.*;
```
`*`을 사용하기보다는 사용하는 것만 import 해보면 좋을 것 같습니다!

hjhello423 commented on src/test/java/nextstep/subway/unit/SectionsTest.java:
```diff
@@ -54,4 +93,57 @@ void setUp() {
+		assertThat(sections.getStations()).containsExactly(신사역, 신논현역);
+	}
+
+	@Test
+	void 지하철_노선에_상행_종점역인_구간을_삭제한다() {
```
단위 테스트도 추가되었네요. 👍

hjhello423 reviewed (approved):
안녕하세요 민규님 😀
리뷰 잘 반영해 주셨네요.
그럼 이번 step은 머지하도록 하겠습니다!
